### PR TITLE
ISPN-12500 Custom server paths

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-12500

### CODE REVIEW

**Comment by tristantarrant** [comment]
@oraNod  this has repercussions on #8880

File: server/runtime/src/main/java/org/infinispan/server/loader/Loader.java [hunk 1]
```diff
@@ -32,6 +34,11 @@
     * directory under the server home.
     */
    public static final String INFINISPAN_SERVER_ROOT_PATH = "infinispan.server.root.path";
+   /**
+    * Property name indicating the paths to the server lib directories. If unspecified, defaults to the <i>lib</i>
+    * directory undert the server root.
```
Comment (oraNod) [inline server/runtime/src/main/java/org/infinispan/server/loader/Loader.java:39]:
```suggestion
    * directory under the server root.
```

File: server/runtime/src/main/server/README.md [hunk 2]
```diff
@@ -89,3 +89,22 @@ The default server configuration requires client authentication. Create a user w
...
+The server has the following directory structure:
+
+* `/bin`: scripts to start the server and the CLI.
+* `/boot`: `JAR` files to boot the server.
+* `/docs`: examples, configuration schemas, component licenses, etc.
+* `/lib`: server `JAR` files. Should not be used for custom jars.
+* `/server`: default server instance folder. Can be overridden with the `-s` or `--server-root` argument.
+- `/server/conf`: configuration files. Can be overridden with the `infinispan.server.conf.path` system property.
+- `/server/data`: data files organized by container name. Can be overridden with the `infinispan.server.data.path` system property.
+- `/server/lib`: server extensions files for custom filters, listeners, and so on. This directory will be scanned recursively and used as a classpath. Can be ovverridden with the `infinispan.server.lib.path` system property which can specify multiple paths separated by the platform path separator (`:` on Unix, `;` on Windows)
```
Comment (oraNod) [inline server/runtime/src/main/server/README.md:106]:
```suggestion
- `/server/lib`: server extensions files for custom filters, listeners, and so on. This directory will be scanned recursively and used as a classpath. Override with the `infinispan.server.lib.path` system property to set a custom location with multiple paths separated by the `:` character on Unix or the `;` character on Windows.
```

**Comment by oraNod** [comment]
@tristantarrant I've updated the README and added a script to pull it into the main docs repo during maven builds. Hope that looks OK.

**Comment by oraNod** [comment]
@tristantarrant Hey, I apologize for all the commits and churn on docs in this PR. Initially I thought we could add a script that replaces things like `${infinispan.brand.name}` with `{brandname}` so the README.adoc gets included in the Server Guide like we do for downstream.

However, we'd need to add the script to fetch_docs.rb for generating the site plus if you build locally outside maven it doesn't work. I think it's best to avoid wiring things together too much with the community docs.

At the moment the docs reference the README in the server distribution, which might not be the best user experience. We could pull the README into the docs directly without replacing variables like `${infinispan.brand.name}`. Do you think we should do that or leave as is? 

Apologies again for faffing about with all the commits here. I normally like to keep things cleaner.

**Comment by tristantarrant** [comment]
No worries. I've fixed up all the commits into one

**Review by oraNod — APPROVED** [review]
Looks good. Thanks @tristantarrant !

File: server/runtime/src/test/java/org/infinispan/server/loader/LoaderTest.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      ShrinkWrap.create(JavaArchive.class,"two.jar")
+            .addClasses(TwoClass.class).as(ZipExporter.class).exportTo(lib2.resolve("two.jar").toFile(), true);
+      ShrinkWrap.create(JavaArchive.class,"three.jar")
+            .addClasses(ThreeClass.class).as(ZipExporter.class).exportTo(lib3.resolve("three.jar").toFile(), true);
+      Properties properties = new Properties();
+      properties.put(Loader.INFINISPAN_SERVER_LIB_PATH, String.join(File.pathSeparator, lib1.toString(), lib2.toString(), lib3.toString()));
+      Loader.run(new String[]{LoaderTest.class.getName()}, properties);
+   }
+
+   public static void main(String args[]) {
```
Comment (pruivo) [inline server/runtime/src/test/java/org/infinispan/server/loader/LoaderTest.java:42]: nit: 
```suggestion
   public static void main(String[] args) {
```

**Review by pruivo — CHANGES_REQUESTED** [review]
minor suggestions :)

File: documentation/src/main/asciidoc/stories/assembly_server_reference.adoc [hunk 4]
```diff
@@ -1,14 +1,10 @@
 [id='server_reference']
-:context: server
+:context: server-info
 = Reference
 
-//Community
```
Comment (pruivo) [inline documentation/src/main/asciidoc/stories/assembly_server_reference.adoc:5]: @oraNod is this intentional? (removing from community)

Comment (pruivo) [reply documentation/src/main/asciidoc/stories/assembly_server_reference.adoc:5]: ps. I think this file isn't consumed anywhere now

Comment (oraNod) [reply documentation/src/main/asciidoc/stories/assembly_server_reference.adoc:5]: @pruivo Yes, this is not included in the community edition of the Server Guide. I do want to leave the assembly in the community doc set. I'm planning to come back to this and figure it out separately from this PR.

File: server/runtime/src/main/server/README.adoc [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+== Logging
+Configure ${infinispan.brand.name} Server logging with the `log4j2.xml` file in the `server/conf` folder.
+
+Use the `--logging-config=<path_to_logfile>` argument to use custom paths, as follows:
+
+.Unix / Linux
+
+[source,options="nowrap",subs=attributes+]
+----
+<SERVER_HOME>/bin/server.sh --logging-config=/.opt/etc/log4j2-w.xml
```
Comment (pruivo) [inline server/runtime/src/main/server/README.adoc:270]:
```suggestion
<SERVER_HOME>/bin/server.sh --logging-config=/path/to/log4j2.xml
```

File: server/runtime/src/main/server/README.adoc [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+----
+
+TIP: To ensure custom paths take effect, do not use the `~` shortcut.
+
+.Windows
+
+[source,options="nowrap",subs=attributes+]
+----
+<SERVER_HOME>\bin\server.bat --logging-config=path\to\log4j2-w.xml
```
Comment (pruivo) [inline server/runtime/src/main/server/README.adoc:279]:
```suggestion
<SERVER_HOME>\bin\server.bat --logging-config=path\to\log4j2.xml
```

**Comment by wfink** [comment]
The infinispan.server.lib.path work if added to JAVA_OPTS but not added with the -P option.
The settings for log/config/data are working for -P

**Comment by tristantarrant** [comment]
because lib is processed by the Loader which has no clue about -P

File: server/runtime/src/main/java/org/infinispan/server/loader/Loader.java [hunk 7]
```diff
@@ -82,21 +90,27 @@ public static void run(String[] args, Properties properties) {
...
-                     } else {
-                        jars.put(artifact, jar.toUri().toURL());
-                     }
-                  } catch (MalformedURLException e) {
+         if (!Files.exists(path)) {
+            return parent;
+         }
+         Map<String, URL> urls = new LinkedHashMap<>();
+         Files.find(path, Integer.MAX_VALUE, (p, a) -> p.toString().endsWith(".jar") || a.isDirectory(), FileVisitOption.FOLLOW_LINKS).forEach(p -> {
+            try {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/loader/Loader.java:98]: Ok

File: server/runtime/src/main/java/org/infinispan/server/loader/Loader.java [hunk 8]
```diff
@@ -82,21 +90,27 @@ public static void run(String[] args, Properties properties) {
...
+                  String artifact = extractArtifactName(p.getFileName().toString());
+                  if (urls.containsKey(artifact)) {
+                     throw new IllegalArgumentException("Duplicate JARs:\n" + p + "\n" + urls.get(artifact));
+                  } else {
+                     urls.put(artifact, p.toUri().toURL());
                   }
-               });
-         final URL[] array = jars.values().toArray(new URL[jars.size()]);
+               } else {
+                  // It's a directory
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/java/org/infinispan/server/loader/Loader.java:107]: `find()` returns both files and directories and we need to process them differently. `URLClassLoader` treats non-jar URIs as root paths where to look for "bare" resources such as property files.

**Comment by pruivo** [comment]
> because lib is processed by the Loader which has no clue about -P

are you going to fix it? `infinispan.server.root.path` is ignored as well.

**Comment by tristantarrant** [comment]
Working on it now

File: server/runtime/src/main/server/README.adoc [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+
+[%header,cols=3*]
+|===
+|Folder
+|Description
+|System property override
+
+|`/server/conf`
+|Contains server configuration files.
+|`infinispan.server.conf.path`
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/runtime/src/main/server/README.adoc]: That's because it's `infinispan.server.config.path`. I've fixed the doc.

**Comment by tristantarrant** [comment]
`Loader` now processes `infinispan.server.home.path`. `infinispan.server.root.path` and `infinispan.server.lib.path` both when specified as system properties as well as passed in through `-P/--properties=`

**Comment by pruivo** [comment]
@tristantarrant what about `infinispan.service`? Are you going to move it to `docs/` or leave it in `bin/systemd`?

**Review by pruivo — APPROVED** [review]

**Comment by tristantarrant** [comment]
@pruivo done

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
integrated! thanks @tristantarrant !

**Pull request merged**